### PR TITLE
Add DEV_MODE to support continuous build&run when changing sources.

### PR DESCRIPTION
**Pull request by tmds**

### CODE REVIEW

**Comment by tmds** [comment]
Implements https://github.com/redhat-developer/s2i-dotnetcore/issues/116

**Comment by tmds** [comment]
CC @mickaelistria

**Comment by mickaelistria** [comment]
So this would re-run `dotnet restore` and then `dotnet run` on any file change?

**Comment by tmds** [comment]
> So this would re-run dotnet restore and then dotnet run on any file change?

Yes. If restoring turns out to be a major source of delay we can see how we can improve. The current implementation is the simplest :smile:

**Comment by omajid** [comment]
`dotnet run --help` mentions both `--no-restore` and `--no-build`. It seems like both are implicit. Could we simply do a `dotnet run` instead of both? There is a number of issues open about slow performance of run (and other SDK commands), so it might also be worth running `dotnet /path/to/dll` directly.

**Comment by omajid** [comment]
Please watch out for https://github.com/dotnet/cli/issues/7583 too.

**Comment by tmds** [comment]
@omajid Thanks for suggestions.

> dotnet run --help mentions both --no-restore and --no-build. It seems like both are implicit. Could we simply do a dotnet run instead of both? There is a number of issues open about slow performance of run (and other SDK commands), so it might also be worth running dotnet /path/to/dll directly.

We are doing the restore separately because it is convenient to pass the DOTNET_RESTORE_SOURCES. We could do it implicit as part of the build when there are no restore sources, but that would introduce perhaps slightly different behavior and an additional code path. We need to build and start the application and this is what dotnet run does. If we do some measurements that show an interesting gain, we can apply improvements in additional PRs.

> Please watch out for dotnet/cli#7583 too.

It seems that issue is related to the working directory. We are cd-ing into DOTNET_STARTUP_PROJECT.

File: 2.0/build/Dockerfile.rhel7 [hunk 1]
```diff
@@ -20,13 +20,18 @@ LABEL name="dotnet/dotnet-20-rhel7" \
...
+
 # Switch to root for package installs
 USER 0
 
 # Copy the S2I scripts from the specific language image to $STI_SCRIPTS_PATH.
 COPY ./s2i/bin/ /usr/libexec/s2i
 
-RUN INSTALL_PKGS="rh-nodejs6-npm rh-dotnet20-dotnet-sdk-2.0 rsync" && \
+RUN INSTALL_PKGS="rh-nodejs6-npm rh-nodejs6-nodejs-nodemon rh-dotnet20-dotnet-sdk-2.0 rsync" && \
+    ln -s /usr/lib/node_modules/nodemon/bin/nodemon.js /usr/bin/nodemon && \
```
Comment (jerboaa) [inline 2.0/build/Dockerfile.rhel7]: This symlink seems to be a non-standard way of doing things. I'd expect for nodemon to be in the PATH when the rh-nodejs6 collection is enabled. Is it not? Could you please check with rh-nodejs6 maintainers? What are they doing for having nodemon in the path?

Comment (tmds) [reply 2.0/build/Dockerfile.rhel7]: I copied this from the s2i nodejs without giving it much attention. Strange indeed. Going to check if it works without the symlink.

Comment (tmds) [reply 2.0/build/Dockerfile.rhel7]: Works fine without the symlink!

**Comment by rgrunber** [comment]
I've built against this change and get the following error in the runtime container when setting DEV_MODE=true : https://paste.fedoraproject.org/paste/9LWSRj5YQ60pW-FndwBYwA . If I don't include DEV_MODE in the environment, then everything works as expected (application starts up), but of course the files modified don't trigger updates to the application. Any advice ?

**Comment by tmds** [comment]
> /opt/rh/rh-dotnet20/root/usr/lib64/dotnet/sdk/2.0.0/NuGet.targets(102,5): error : Access to the path '/opt/app-root/src/obj/262212f2-75b0-4405-ab91-62440fb1b01b.tmp' is denied. [/opt/app-root/src/NewDotnetProject.csproj]
/opt/rh/rh-dotnet20/root/usr/lib64/dotnet/sdk/2.0.0/NuGet.targets(102,5): error :   Permission denied [/opt/app-root/src/NewDotnetProject.csproj]

@rgrunber I've pushed a change that _may_ fix this. Please give this a try. If it still fails, can you share the permissions&ownership of the file reported in the error message (filename will change for each build) and it's parent folder ('obj')?

**Comment by jerboaa** [comment]
@tmds Would we be able to expand the test for this issue. It sounds like it would be reproducible if the build-container and copy steps are done with different UIDs. Thoughts?

**Comment by tmds** [comment]
I've added a test. Without fix-permission it fails with a permission error on `obj`. This is the first thing dotnet tries to write.
Probably @rgrunber 's repo has the `obj` folder checked in, so it fails on the second thing dotnet tries to write (`obj/<guid>.tmp`).

**Comment by rgrunber** [comment]
@tmds Yes, this seems to have been the case. I'll test the currrent changes but most likely it'll work now.

**Comment by rgrunber** [comment]
This works for me. A simple refresh (10-15s after modification) show the updated content. Thanks again for the help!

**Comment by tmds** [comment]
I feel comfortable merging this now 😄 
Thanks for reviewing and smoke testing @jerboaa @omajid @rgrunber

**Pull request merged**